core/src/OpenMP/Kokkos_OpenMP_Instance.cpp

```diff
@@ -139,13 +141,19 @@ void OpenMPInternal::resize_thread_data(size_t pool_reduce_bytes,
   }
 }
 
+OpenMPInternal *OpenMPInternal::s_singleton = nullptr;
+
 OpenMPInternal &OpenMPInternal::singleton() {
-  static OpenMPInternal *self = nullptr;
-  if (self == nullptr) {
-    self = new OpenMPInternal(get_current_max_threads());
-  }
+  allocate_singleton();
+  return *s_singleton;
+}
+
+void OpenMPInternal::allocate_singleton() {
+  if (!s_singleton) s_singleton = new OpenMPInternal(get_current_max_threads());
+}
 
-  return *self;
+void OpenMPInternal::deallocate_singleton() {
+  if (s_singleton) delete s_singleton;
 }
```

Comment:
Shouldn't be public. Just check in `finalize` if the current instance is the singleton and delete it in that case.

Reply:
Finalize is not a static member function. I'm not sure it is safe to delete `this` in `finalize`.

Reply:
We decided to use `finalize` to call `clear_thread_data`.